src/catchup/CatchupManagerImpl.cpp

```diff
@@ -28,32 +28,132 @@ CatchupManager::create(Application& app)
 }
 
 CatchupManagerImpl::CatchupManagerImpl(Application& app)
-    : mApp(app), mCatchupWork(nullptr)
+    : mApp(app)
+    , mCatchupWork(nullptr)
+    , mSyncingLedgersSize(
+          app.getMetrics().NewCounter({"ledger", "memory", "queued-ledgers"}))
 {
 }
 
 CatchupManagerImpl::~CatchupManagerImpl()
 {
 }
 
+uint32_t
+CatchupManagerImpl::getCatchupCount()
+{
+    return mApp.getConfig().CATCHUP_COMPLETE
+               ? std::numeric_limits<uint32_t>::max()
+               : mApp.getConfig().CATCHUP_RECENT;
+}
+
 void
-CatchupManagerImpl::historyCaughtup()
+CatchupManagerImpl::processLedger(LedgerCloseData const& ledgerData)
 {
-    mCatchupWork.reset();
+    uint32_t lastReceivedLedgerSeq = ledgerData.getLedgerSeq();
+
+    // Caught up
+    if (lastReceivedLedgerSeq ==
+        mApp.getLedgerManager().getLastClosedLedgerNum())
+    {
+        reset();
+        return;
+    }
+
+    // We are not caught up, but catchup work is finished
+    if (mCatchupWork && mCatchupWork->isDone())
+    {
+        trimSyncingLedgers();
+        mCatchupWork.reset();
+    }
+
+    auto addLedgerData = [&]() { addToSyncingLedgers(ledgerData); };
+
+    if (!mCatchupWork)
+    {
+        uint32_t checkpointLedger;
+        if (!mSyncingLedgers.empty())
+        {
+            // front of mSyncingLedgers should always be a checkpoint ledger if
+            // catchup is not running
+            uint32_t firstBufferedLedgerSeq =
+                mSyncingLedgers.front().getLedgerSeq();
+
+            checkpointLedger = mApp.getHistoryManager().nextCheckpointLedger(
+                firstBufferedLedgerSeq - 1);
+            assert(checkpointLedger == firstBufferedLedgerSeq);
+        }
+        else
+        {
+            checkpointLedger = mApp.getHistoryManager().nextCheckpointLedger(
+                lastReceivedLedgerSeq - 1);
+
+            if (lastReceivedLedgerSeq == checkpointLedger)
+            {
+                // We wait until checkpoint + 1 ledger is closed before we start
+                // catchup, but we want to catchup to checkpoint - 1 so we don't
+                // have to wait for the next checkpoint snapshot. We need to
+                // buffer the checkpoint ledger because of this.
+                addLedgerData();
+            }
+            else if (lastReceivedLedgerSeq == checkpointLedger + 1)
+            {
+                // if first ledger received is right after a checkpoint, but the
+                // checkpoint was not buffered, then we need to wait for the
+                // next checkpoint.
+                checkpointLedger =
+                    mApp.getHistoryManager().nextCheckpointLedger(
+                        lastReceivedLedgerSeq);
+            }
+        }
+
+        std::string message;
+        uint32_t catchupTriggerLedger = checkpointLedger + 1;
+        if (lastReceivedLedgerSeq >= catchupTriggerLedger)
```

Comment:
Catchup can be triggered when `lastReceivedLedgerSeq > catchupTriggerLedger` if catchup fails and we still have syncing ledgers after trimming.

Reply:
Yeah, sorry, I didn't see that we're now checking ledger validity in `ApplyBufferedLedgers` (which will allow catchup to properly recover and not crash), so that should be fine!